File: 0606. Construct String from Binary Tree/Solution.cpp
// Author        : tarlansoltanov
// Title         : Construct String from Binary Tree
// ID            : 0606
// URL           : https://leetcode.com/problems/construct-string-from-binary-tree/
// Difficulty    : Easy
// Language      : C++


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    string tree2str(TreeNode* root) {
        string res = to_string(root->val);
        if(root->left == nullptr && root->right == nullptr) return res;
        if(root->left == nullptr) return res + "()" + "(" + tree2str(root->right) + ")";
        if(root->right == nullptr) return res + "(" + tree2str(root->left) + ")";
        return res + "(" + tree2str(root->left) + ")" + "(" + tree2str(root->right) + ")";
    }
};

File: 0606. Construct String from Binary Tree/Solution.py
# Author        : tarlansoltanov
# Title         : Construct String from Binary Tree
# ID            : 0606
# URL           : https://leetcode.com/problems/construct-string-from-binary-tree/
# Difficulty    : Easy
# Language      : Python


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def tree2str(self, root: Optional[TreeNode]) -> str:
        if root.left is None and root.right is None:
            return f"{root.val}"
        return f"{root.val}{f'({self.tree2str(root.left)})' if root.left else '()'}{f'({self.tree2str(root.right)})' if root.right else ''}"
        